src/uct/ib/efa/ib_efa_md.c

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <uct/ib/efa/ib_efa.h>
+#include <uct/ib/base/ib_md.h>
+#include <ucs/type/status.h>
+#include <ucs/debug/log.h>
+
+
+typedef struct uct_ib_efa_mem {
+    uct_ib_mem_t        super;
+    uct_ib_mr_t         mrs[];
+} uct_ib_efa_mem_t;
+
+static uint64_t
+uct_ib_efadv_access_flags(const uct_ib_efadv_t *efadv)
+{
+    uint64_t access_flags = IBV_ACCESS_LOCAL_WRITE;
+#ifdef HAVE_DECL_EFA_DV_RDMA_READ
+    if (efadv->efadv_attr.device_caps & EFADV_DEVICE_ATTR_CAPS_RDMA_READ) {
+        access_flags |= IBV_ACCESS_REMOTE_READ;
+    }
+#endif
+    return access_flags;
+}
+
+static ucs_status_t uct_ib_efa_reg_key(uct_ib_md_t *md, void *address,
+                                       size_t length, uint64_t access_flags,
+                                       uct_ib_mem_t *ib_memh,
+                                       uct_ib_mr_type_t mr_type, int silent)
+{
+    uct_ib_efa_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_efa_mem_t);
+
+    return uct_ib_reg_key_impl(md, address, length, access_flags,
+                               ib_memh, &memh->mrs[mr_type], mr_type, silent);
+}
+
+static ucs_status_t uct_ib_efa_dereg_key(uct_ib_md_t *md,
+                                         uct_ib_mem_t *ib_memh,
+                                         uct_ib_mr_type_t mr_type)
+{
+    uct_ib_efa_mem_t *memh = ucs_derived_of(ib_memh, uct_ib_efa_mem_t);
+
+    return uct_ib_dereg_mr(memh->mrs[mr_type].ib);
+}
```

Comment:
do we plan to extend these functions for EFA?

Reply:
Not really. So, should we reuse `uct_ib_verbs_reg_key()` and `uct_ib_verbs_dereg_key()` and `uct_ib_verbs_mem_t` for EFA md?
I think this will be also related to your comment regarding common code with verbs.